accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,290 @@
+
+For example, given the following content in the `package.json`:
+```json
+ "devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.3.2"
+  }
+```
+
+The `yarn.lock` file will contain:
+```
+"typescript@>=2.0.0 <2.3.0":
+  version "2.2.2"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-2.2.2.tgz#606022508479b55ffa368b58fee963a03dfd7b0c"
+
+typescript@2.3.2:
+  version "2.3.2"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-2.3.2.tgz#f0f045e196f69a72f06b25fd3bd39d01c3ce9984"
+```
+
+Also, there will be:
+- `typescript@2.3.2` in `node_modules/typescript`
+- `typescript@2.2.2` in `node_modules/@angular/cli/node_modules`.
+
+## Problem
+
+In this context, it is impossible to force the use of `typescript@2.3.2` for
+the whole project (except by flattening the whole project, which we don't want).
+
+It makes sense for typescript as the user intent is clearly to use typescript
+2.3.2 for compiling all its project, and with the current behaviour, the angular
+CLI (responsible of compiling `.ts` files) will simply use the 2.2.2 version
+from its `node_modules`.
+
+## Variations of the original problem
+
+Similarly, even using such a content for `package.json`:
+```json
+ "devDependencies": {
+    "@angular/cli": "1.0.3"
+  }
+```
+
+The need could arise for forcing the use of `typescript@2.3.2` (or
+`typescript@2.1.0` for that matter).
+
+In these example, the need does not seem very important (the user could maybe 
+use `typescript@2.2.2` or ask the `@angular/cli` dev team to relax its
+constraints on typescript), but there could be cases where a nested dependency
+introduces a bug and the project developer would want to set a specific
+version for it (see for example this
+[comment](https://github.com/yarnpkg/yarn/issues/2763#issuecomment-302682844)).
+
+## Related scenario (out of scope of this document)
+
+An extension of this motivation is also the potential need for mapping nested
+dependencies to others. For example a project developer could want to map
+`typescript@>=2.0.0 <2.3.0` to `my-typescript-fork@2.0.0`.
+
+See alternatives solutions below also.
+
+# Detailed design
+
+The proposed solution is to make the `resolutions` field of the `package.json`
+file to be considered all the time and on a per-package basis.
+
+When a nested dependency is being resolved by yarn, if the `resolutions` field
+contains a version for this package, then this version is used instead.
+
+Most of the examples are given with exact dependencies, but note that putting a
+non-exact specification in the `resolutions` field should be accepted and
+resolved by yarn like it usually does. This subject is discussed below also.
+
+Any potentially counter-intuitive situation will result in a warning being
+issued. This subject is discussed at the end of this section.
+
+## Example
+
+For example with:
+```json
+ "devDependencies": {
+    "@angular/cli": "1.0.3"
+  },
+  "resolutions": {
+    "typescript": "2.3.2"

```

Comment:
I still think it is a bit too brute force to force typescript to `2.3.2` everywhere.
We need to be able to enforce typescript to some subtree.
What do you think of the proposed glob for the resolution path?

The use case I am trying to solve is this:
- left-pad@2.0.5 is getting released with a bug
- it is used all over the app dependencies, some of them get locked on @1.0.1 which is fine 
- you want to upgrade jest to next version but some of the subdependencies requires "leftpad@^2.0.0" which will get resolved to "leftpad@2.0.5"
- you want to force jest's subdependency to leftpad@2.0.4 but don't change force it across the whole app

I would think this could solve it elegantly.
```
"resolutions": {
  "jest-*/**/left-pad": "1.0.4"
}
```

What do you think?